repose-aggregator/components/filters/openstack-identity-v3-filter/src/main/scala/org/openrepose/filters/openstackidentityv3/utilities/Cache.scala

```diff
@@ -0,0 +1,78 @@
+/*
+ * _=_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_=
+ * Repose
+ * _-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-
+ * Copyright (C) 2010 - 2015 Rackspace US, Inc.
+ * _-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * =_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_=_
+ */
+package org.openrepose.filters.openstackidentityv3.utilities
+
+import java.io.Serializable
+import java.util.concurrent.TimeUnit
+
+import org.openrepose.core.services.datastore.{Datastore, Patch}
+
+import scala.util.Random
+
+object Cache {
+  final val AdminTokenKey = "IDENTITY:V3:ADMIN_TOKEN"
+  final val TokenKeyPrefix = "IDENTITY:V3:TOKEN:"
+  final val GroupsKeyPrefix = "IDENTITY:V3:GROUPS:"
+  final val UserIdKeyPrefix = "IDENTITY:V3:USER_ID:"
+
+  def getTokenKey(token: String): String = TokenKeyPrefix + token
+
+  def getGroupsKey(token: String): String = GroupsKeyPrefix + token
+
+  def getUserIdKey(userId: String): String = UserIdKeyPrefix + userId
+
+  def offsetTtl(exactTtl: Int, offset: Int): Int = {
+    if (offset == 0 || exactTtl == 0) exactTtl
+    else safeLongToInt(exactTtl.toLong + (Random.nextInt(offset * 2) - offset))
+  }
+
+  def safeLongToInt(long: Long): Int = {
+    if (long < 0) math.max(long, Int.MinValue).toInt
+    else math.min(long, Int.MaxValue).toInt
+  }
+}
+
+class Cache(datastore: Datastore) extends Datastore {
```

Comment:
Trying to organize code, and extract caching related stuff to a more sensible place.

Or, for this class more specifically, to add the `safeGet` method. I suppose this is kind of a verbose way of doing that. I could make it a static function, it just requires another parameter in this parameter list.